File: src/com/checkers/board/Man.java
package com.checkers.board;

import com.checkers.exceptions.*;
import com.checkers.utils.Colour;

public class Man extends Piece {

    public Man(Colour colour) {
        setColour(colour);
    }

    public void analyzeAbilityOfMove() throws CheckersException {
        boolean firstCell = false;
        boolean secondCell = false;
        if (getColour() == Colour.WHITE) {
            try {
                firstCell = isAbleToMoveTo(getCell().getNear(1, -1));
            } catch (ArrayIndexOutOfBoundsException ex) {}

            try {
                secondCell = isAbleToMoveTo(getCell().getNear(1, 1));
            } catch (ArrayIndexOutOfBoundsException ex) {}

            if (firstCell || secondCell) {
                setCanMove(true);
            } else {
                setCanMove(false);
            }
        } else {
            try {
                firstCell = isAbleToMoveTo(getCell().getNear(-1, -1));
            } catch (ArrayIndexOutOfBoundsException ex) {}

            try {
                secondCell = isAbleToMoveTo(getCell().getNear(-1, 1));
            } catch (ArrayIndexOutOfBoundsException ex) {}

            if (firstCell || secondCell) {
                setCanMove(true);
            } else {
                setCanMove(false);
            }
        }
    }

    public void analyzeAbilityOfEat() throws CheckersException {
        Cell pieceCell = getCell();
        boolean first = false;
        boolean second = false;
        boolean third = false;
        boolean fourth = false;

        try {
            first = isAbleToEatTo(pieceCell.getNear(2, 2));
        } catch (ArrayIndexOutOfBoundsException ex) {}

        try {
            second = isAbleToEatTo(pieceCell.getNear(-2, 2));
        } catch (ArrayIndexOutOfBoundsException ex) {}

        try {
            third = isAbleToEatTo(pieceCell.getNear(-2, -2));
        } catch (ArrayIndexOutOfBoundsException ex) {}

        try {
            fourth = isAbleToEatTo(pieceCell.getNear(2, -2));
        } catch (ArrayIndexOutOfBoundsException ex) {}

        if (first || second || third || fourth) setCanEat(true);
        else setCanEat(false);
    }

    public boolean isAbleToMoveTo(Cell to) throws CheckersException {
        Cell pieceCell = getCell();
        if (pieceCell.diff(to) != 1) return false;
        boolean firstCell = false;
        boolean secondCell = false;
        if (getColour() == Colour.WHITE) {
            try {
                firstCell = pieceCell.getNear(1, 1) == to;
            } catch (ArrayIndexOutOfBoundsException ex) {}

            try {
                secondCell = pieceCell.getNear(1, -1) == to;
            } catch (ArrayIndexOutOfBoundsException ex) {}

            if (firstCell || secondCell) return true;
        } else {
            try {
                firstCell = pieceCell.getNear(-1, -1) == to;
            } catch (ArrayIndexOutOfBoundsException ex) {}

            try {
                Cell cell = pieceCell.getNear(-1, 1);
                secondCell = cell == to;
            } catch (ArrayIndexOutOfBoundsException ex) {}

            if (firstCell || secondCell) return true;
        }
        return false;
    }

    public boolean isAbleToEatTo(Cell to) throws CheckersException {
        Cell pieceCell = getCell();
        if (pieceCell.diff(to) != 2) return false;
        else if (to.getPiece() != null) return false;
        Cell target = pieceCell.between(to, pieceCell.getBoard());
        if (target.getPiece() == null) return false;
        if (target.getPiece().getColour() == getColour()) return false;
        return true;
    }

    public void move(Cell to) throws CheckersException {
        if (isCanEat()) throw new CanEatException();
        if (!isCanMove()) throw new CanNotMoveException();
        if (!isAbleToMoveTo(to)) throw new CanNotMoveException();
        Cell from = getCell();
        to.setPiece(this);
        from.setPiece(null);
    }

    public void eat(Cell to) throws CheckersException {
        if (!isCanEat()) throw new CanNotEatException();
        if (!isAbleToEatTo(to)) throw new CanNotEatException();
        Cell from = getCell();
        to.setPiece(this);
        from.setPiece(null);
        from.between(to, from.getBoard()).setPiece(null);
    }

    @Override
    public String toString() {
        if (getColour() == Colour.BLACK) {
            return "@";
        } else {
            return "$";
        }
    }
}

File: src/main/kotlin/com/checkers/board/Man.kt
package com.checkers.board

import com.checkers.exception.CannotEatException
import com.checkers.exception.CannotMoveException
import com.checkers.exception.MustEatException
import com.checkers.utils.Color
import com.checkers.utils.isCoordinateExists
import com.checkers.utils.isCoordinatesExists

class Man(color: Color) : Piece(color) {

    override fun analyzeAbilityOfMove() {
        val nextRow = getNextRow()
        if (!isCoordinateExists(nextRow)) {
            return
        }

        val board = cell.board
        val column = cell.column
        isCanMove = sequenceOf(column - 1, column + 1)
            .filter { isCoordinateExists(it) }
            .map { board.getCell(nextRow, it) }
            .any { isAbleToMoveTo(it) }
    }

    private fun getNextRow(): Int =
        if (color == Color.WHITE) {
            cell.row + 1
        } else {
            cell.row - 1
        }

    override fun analyzeAbilityOfEat() {
        val column = cell.column
        val row = cell.row
        val board = cell.board
        isCanEat = sequenceOf(row - 2, row + 2)
            .filter { isCoordinateExists(it) }
            .flatMap { nextRow ->
                sequenceOf(column - 2, column + 2)
                    .filter { isCoordinateExists(it) }
                    .map { board.getCell(nextRow, it) }
            }
            .any { isAbleToEatTo(it) }
    }

    override fun isAbleToMoveTo(destinationCell: Cell): Boolean {
        if (destinationCell.piece != null || isEnemyNear() || cell.diff(destinationCell) != 1) {
            return false
        }

        val nextRow = getNextRow()
        if (!isCoordinateExists(nextRow)) {
            return false
        }

        val column = cell.column
        val board = cell.board
        return sequenceOf(column - 1, column + 1)
            .filter { isCoordinateExists(it) }
            .map { board.getCell(nextRow, it) }
            .any { it == destinationCell }
    }

    private fun isEnemyNear(): Boolean =
        sequenceOf(
            cell.row + 1 to cell.column + 1,
            cell.row + 1 to cell.column - 1,
            cell.row - 1 to cell.column - 1,
            cell.row - 1 to cell.column + 1,
        )
            .filter { isCoordinatesExists(it.first, it.second) }
            .map { cell.board.getCell(it.first, it.second) }
            .any { hasEnemyIn(it) }

    private fun hasEnemyIn(cell: Cell): Boolean =
        cell.piece?.let { it.color != color } ?: false

    override fun isAbleToEatTo(destinationCell: Cell): Boolean {
        if (cell.diff(destinationCell) != 2 || destinationCell.piece != null) {
            return false
        }

        return cell.between(destinationCell, cell.board)
            .piece?.let { it.color != color } ?: false
    }

    override fun move(destinationCell: Cell) {
        if (isCanEat) {
            throw MustEatException()
        }

        if (!isCanMove) {
            throw CannotMoveException(cell, destinationCell)
        }

        if (!isAbleToMoveTo(destinationCell)) {
            throw CannotMoveException(cell, destinationCell)
        }

        cell.piece = null
        updatePiece(destinationCell)
    }

    override fun eat(destinationCell: Cell) {
        if (!isCanEat || !isAbleToEatTo(destinationCell)) {
            throw CannotEatException(cell, destinationCell)
        }

        cell.piece = null
        cell.between(destinationCell, cell.board).piece = null
        updatePiece(destinationCell)
    }

    private fun updatePiece(cell: Cell) =
        if (color == Color.WHITE && cell.row == 8 || color == Color.BLACK && cell.row == 1) {
            cell.piece = createKing()
        } else {
            cell.piece = this
        }

    private fun createKing(): Piece =
        King(color)

    override fun isMan(): Boolean =
        true

    override fun isKing(): Boolean =
        false
}